File: Arrays/62-unique-paths/unique-paths.java
class Solution {
    public int uniquePaths(int m, int n) {
        // defining a dp matrix and intializing all cells as -1;
        int dp[][] = new int[m][n];
        for(int i=0; i<dp.length;i++){
            for(int j=0;j<dp[0].length;j++){
                dp[i][j]=-1;
            }
        }
        return result(0,0,m,n, dp);
    }

    static int result(int row, int col, int m, int n, int[][] dp){

        //return 1 if we reach the end of the array
        if(row==m-1 && col==n-1){
            return 1;
        }

        //return the cell value if we already have traveresed it
        if(dp[row][col]!=-1){
            return dp[row][col];
        }

        //memoization: If the path has not been traversed yet
        if(row == m-1){
            return dp[row][col]=result(row, col+1, m, n, dp);
        }else if(col == n-1){
           return dp[row][col]=result(row+1, col, m, n, dp);
        }

        return dp[row][col]=result(row+1, col, m, n, dp)+result(row, col+1, m, n, dp);
    }
}

File: DP/62-unique-paths/unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        
        rows = m
        cols = n

        #Memoization
        # dp = [ [-1 for _ in range(cols)] for _ in range(rows)]

        # print(dp)


        # def rec(row, col):
        #     if row == 0 or col == 0:
        #         return 1

        #     if dp[row][col] != -1:
        #         return dp[row][col]

        #     up = rec(row-1, col)
        #     left = rec(row, col - 1)

        #     dp[row][col] =  up + left
        #     return dp[row][col]

        # return rec(rows-1, cols-1)

        #Tabulation

        dp = [ [-1 for _ in range(cols)] for _ in range(rows)]

        for i in range(rows):
            for j in range(cols):
                if i == 0 or j == 0:
                    dp[i][j] = 1

        for i in range(1, rows):
            for j in range(1, cols):
                dp[i][j] = dp[i-1][j] + dp[i][j-1]

        return dp[rows-1][cols-1]
